hoomd/hpmc/IntegratorHPMCMonoNEC.h

```diff
@@ -0,0 +1,1368 @@
+
+                            //
+                            //  Update Velocities when actually colliding
+                            //  otherwise they will collide again in the next step.
+                            //
+                            vel_n -= vel_change;
+                            vel_k += vel_change;
+
+                            h_velocities.data[next] = make_scalar4( vel_n.x, vel_n.y, vel_n.z, h_velocities.data[next].w);
+                            h_velocities.data[k]    = make_scalar4( vel_k.x, vel_k.y, vel_k.z, h_velocities.data[k].w);
+
+                            velocity = sqrt( dot(vel_n,vel_n));
+
+                            direction = vel_n / velocity;
+                        #ifdef ENABLE_MPI
+                            }
+                        else
+                            { // if colliding with an inactive particle.
+                                vec3<Scalar> vel_change = collisionPlaneVector * (dot(vel_k,collisionPlaneVector) / dot(collisionPlaneVector,collisionPlaneVector));
+                                vel_k = vel_k - 2.0 * vel_change;
+
+                                h_velocities.data[k]    = make_scalar4( vel_k.x, vel_k.y, vel_k.z, h_velocities.data[k].w);
+                                next = k;
+
+                                velocity = sqrt( dot(vel_k,vel_k));
+
+                                direction = vel_k / velocity;
+                            }
+                        #endif
+
+                        if( velocity == 0.0 )
+                            {
+                            this->m_exec_conf->msg->warning() << "Cannot continue a chain without moving.\n";
+                            next = -1;
+                            }
+
+                            // store previous particle.
+                            //prev = k;
+                        }
+                    } // end loop over totalDist.
+
+//                 counters.translate_accept_count++;
+
+                }
+            else
+                {
+                // Get the RNG for current particle
+                hoomd::RandomGenerator rng_i(hoomd::Seed(hoomd::RNGIdentifier::HPMCMonoTrialMove, timestep, seed),
+                                             hoomd::Counter(i, this->m_exec_conf->getRank(), i_nselect));
+
+                bool overlap = false;
+                Scalar4 postype_i     = h_postype.data[i];
+                Scalar4 orientation_i = h_orientation.data[i];
+                vec3<Scalar> pos_i    = vec3<Scalar>(postype_i);
+                Shape shape_old(quat<Scalar>(orientation_i), this->m_params[typ_i]);
+
+
+                //move_rotate(shape_i.orientation, rng_i, h_a.data[typ_i], ndim);
+                if (ndim == 2)
+                    move_rotate<2>(shape_i.orientation, rng_i, h_a.data[typ_i]);
+                else
+                    move_rotate<3>(shape_i.orientation, rng_i, h_a.data[typ_i]);
```

Comment:
```suggestion
                move_rotate<ndim>(shape_i.orientation, rng_i, h_a.data[typ_i]);
```

Reply:
with that change I get an error:
`error: the value of ‘ndim’ is not usable in a constant expression`

Reply:
Yeah, the template has to be evaluated at compile time, so something like you had @maklem is necessary.